### DISCUSSION TITLE
Installing VIPS in the context of Docker

### DISCUSSION THREAD

**Discussion by xamir82** [libvips/libvips#3504] [General]
Hi!
I'm building an app that uses VIPS's command-line tools under the hood to do some image manipulation, and I'm currently trying to write a Dockerfile for said app. I'll need VIPS alongside the app itself in final image, but I'm struggling with installing VIPS.

I'm using Ubuntu as the base image in the Dockerfile, so I can technically use `apt` to install VIPS but I'd much prefer to have the latest version instead of the usually-outdated versions available in the default Ubuntu repository.
My first question would be this: Isn't there some sort of a PPA that provides the latest version of VIPS that users can simply use by doing `add-apt-repsoitory`? Much like the PPAs you'd find for FFmpeg, for example.

On my local machine, I've installed VIPS via Homebrew, which has worked perfectly. So the thought of using Homebrew in the Dockerfile did cross my mind, but that would be a pretty unconventional thing to do, I think, (I've never actually seen someone use Homebrew in Docker for this type of use case), so I wanted to ask if you'd recommend that or not?
And I'm not even quite sure how to exactly do it. Should I, for example, install VIPS via Homebrew in one build stage in the Dockerfile and then just copy over the artifacts from `/home/linuxbrew/.linuxbrew` over to the final stage, or what?

I've also considered building from source but it's difficult to the build VIPS from source **with all the ideal dependencies**. For example, `mozjpeg` is apparently recommended but it seems like I would need to build **that** from source as well, in addition to any other dependencies that might not be available via `apt`. The Homebrew version, on the other hand, seems to have been built with the ideal dependencies already.

Also, doesn't VIPS provide static builds somewhere? Once again to use FFmpeg as an example, something like [this](https://ffmpeg.org/download.html#:~:text=Enterprise%20Linux%20packages-,Linux%20Static%20Builds,-32%2Dbit%20and)?

Thank you very much in advance :)

**Comment by jcupitt** [comment]
Hi @xamir82,

We try to spend most of our time on the core library and leave packaging to downstream projects. It takes so much effort to support every platform, and for platforms we don't use ourselves, it's hard to find motivation. I know remi has nice libvips RPMs for redhat-style systems, I think there are PPAs around as well.

I would build from source myself, it's not difficult. I have some sample dockerfiles here:

https://github.com/jcupitt/docker-builds

I would pick one that's similar to what you want and tweak it a bit.

mozjpeg is slightly tricky: libjpeg is one of the core libraries on most systems, and mozjpeg is not binary compatible with the usual IJG libjpeg. This means you also need to build every dependency that uses libjpeg, either directly or indirectly. Unless you have mozjpeg as a hard requirement, I would stick with the standard (and very good) libjpeg-turbo.

**Reply by kengru** [reply]
Thank you, your Debian example worked perfectly for me!

**Comment by jcupitt** [comment]
I made you a dockerfile for mozjpeg:

https://github.com/jcupitt/docker-builds/blob/master/libvips-mozjpeg-ubuntu20.04/Dockerfile

You'd need to add some more useful dependencies though.

**Comment by nickchomey** [comment]
I'm also curious about this - it would be fantastic if there were a ppa repo for ubuntu at the very least. Any ideas on where we might find something like this, or what might be involved in setting it up?

**Reply by jcupitt** [reply]
I think you'd take the current Debian package builder for libvips and backport it to the ubuntu you want to target.

https://launchpad.net/ubuntu/+source/vips/8.15.1-1

Then how you distribute the `.deb` is up to you. A PPA is pretty easy to set up (I think).

**Reply by nickchomey** [reply]
The guy who runs this ppa repo was kind enough to add libvips to it 

https://deb.myguard.nl/packages/

It seems to miss some dependencies and not upgrade others, but it might be of use to some people. 

Some more discussion on that here: https://deb.myguard.nl/forums/topic/request-for-vips-package/#post-4347

**Reply by jcupitt** [reply]
Oh that's nice. I added a comment on updating the jammy package.

**Reply by nickchomey** [reply]
Fantastic, thanks! I followed up with a question about dependency versions. 

Though it looks like the site's forum requires comment approval. It also doesn't seem to provide any notifications.

The main question - does it matter to libvips what version of its dependencies are available?

For example, here's packages installed in jammy and the upcoming noble. 

https://packages.ubuntu.com/jammy/libvips42

https://packages.ubuntu.com/noble/libvips42

But if we add that ppa repo, it installs some versions of noble and some of jammy. In particular, things like libheif and libwebp remain older.

Will libvips work like that or does each version truly have unique minimum dependency versions?

**Reply by jcupitt** [reply]
libvips is pretty flexible. There's a minimum version for each package (enforced by meson), but as long as the package meets that minimum standard, it should all work. If a package is too old, libvips will automatically remove support. The end of meson output includes a useful summary table, eg.:

```
$ meson setup build --prefix ~/vips
The Meson build system
Version: 1.2.1
Source dir: /home/john/GIT/libvips
...
Program python3 found: YES (/usr/bin/python3)
Configuring variables.sh using configuration
Library FuzzingEngine found: NO
Build targets in project: 56
NOTICE: Future-deprecated features used:
 * 0.64.0: {'copy arg in configure_file'}

vips 8.16.0

  Dependencies
    glib-2.0                          : 2.78.0
    gio-2.0                           : 2.78.0
    gobject-2.0                       : 2.78.0
    gmodule-no-export-2.0             : 2.78.0
    expat                             : 2.5.0
    zlib                              : 1.2.13
    libarchive                        : 3.6.2
    fftw3                             : 3.3.10
    cfitsio                           : 4.2.0
    imagequant                        : 2.17.0
    cgif                              : 0.3.2
    libexif                           : 0.6.24
    libjpeg                           : 2.1.5
    libpng                            : 1.6.40
    libwebp                           : 1.2.4
    libwebpmux                        : 1.2.4
    libwebpdemux                      : 1.2.4
    pangocairo                        : 1.51.0
    pangoft2                          : 1.51.0
    fontconfig                        : 2.14.2
    libtiff-4                         : 4.5.1
    librsvg-2.0                       : 2.54.7
    cairo                             : 1.18.0
    matio                             : 1.5.23
    lcms2                             : 2.14
    OpenEXR                           : 3.1.5
    libopenjp2                        : 2.5.0
    libhwy                            : 1.0.7
    niftiio                           : 3.0.1

  Build options
    enable debug                      : NO
    enable deprecated                 : YES
    enable modules                    : YES
    enable gtk-doc                    : NO
    enable doxygen                    : NO
    enable introspection              : YES
    enable examples                   : YES
    enable cplusplus                  : YES
    enable RAD load/save              : YES
    enable Analyze7 load/save         : YES
    enable PPM load/save              : YES
    enable GIF load                   : YES

  Optional external packages
    FFTs with fftw3                   : YES
    SIMD support with libhwy          : YES
    ICC profile support with lcms2    : YES
    deflate compression with zlib     : YES
    text rendering with pangocairo    : YES
    font file support with fontconfig : YES
    EXIF metadata support with libexif: YES

  External image format libraries
    JPEG load/save with libjpeg       : YES
    JXL load/save with libjxl         : YES (dynamic module: YES)
    JPEG2000 load/save with libopenjp2: YES
    PNG load/save with libpng         : YES
    image quantisation with imagequant: YES
    TIFF load/save with libtiff-4     : YES
    image pyramid save with libarchive: YES
    HEIC/AVIF load/save with libheif  : YES (dynamic module: YES)
    WebP load/save with libwebp       : YES
    PDF load with poppler-glib        : YES (dynamic module: YES)
    SVG load with librsvg-2.0         : YES
    EXR load with OpenEXR             : YES
    WSI load with openslide           : YES
    Matlab load with matio            : YES
    NIfTI load/save with niftiio      : YES
    FITS load/save with cfitsio       : YES
    GIF save with cgif                : YES
    Magick load/save with MagickCore  : YES (dynamic module: YES)

  User defined options
    buildtype                         : release
    libdir                            : lib
    prefix                            : /home/john/vips

Found ninja-1.11.1 at /usr/bin/ninja
$ 
```

The downside of course is that you may be missing useful bugfixes.

**Reply by nickchomey** [reply]
Thanks. I'll wait til noble is released, which should alleviate these issues for a while. But ultimately will probably need to do a custom build to keep things up to date. 

Is there any way to make a static build of libvips that includes all (or specified) dependencies?

**Reply by jcupitt** [reply]
You can do eg.:

```
$ meson setup build --prefix ~/vips --default-library static
...
$ cd build
$ ninja install
$ ls -l ~/vips/lib/x86_64-linux-gnu/
total 7192
drwxr-xr-x 2 john john    4096 Jan 28 22:21 girepository-1.0
-rw-r--r-- 1 john john 6772808 Jan 28 22:21 libvips.a
-rw-r--r-- 1 john john  574414 Jan 28 22:21 libvips-cpp.a
drwxr-xr-x 2 john john    4096 Jan 28 22:21 pkgconfig
drwxr-xr-x 2 john john    4096 Jan 28 22:21 vips-modules-8.16
```

But you'll also need to have static libraries for all dependencies, which is where it gets tricky. You'll find `ldd ~/vips/bin/vips` still has a looooong list of stuff.

Deploying in a docker container which has an up to date build maintained by someone else is usually a lot simpler!

**Reply by nickchomey** [reply]
Thanks. Seems like Docker is the way forward for all of this!
